baybe/surrogates/base.py

```diff
@@ -124,60 +134,80 @@ def _make_target_scaler() -> OutcomeTransform | None:
-
-        # Create the composite scaler from the parameter-wise scaler objects
-        transformers = [
-            (
-                "passthrough" if (s := self._make_parameter_scaler(p)) is None else s,
-                [c for c in p.comp_rep_columns if c in searchspace.comp_rep_columns],
-            )
-            for p in searchspace.parameters
-        ]
-        scaler = make_column_transformer(*transformers, verbose_feature_names_out=False)
-
-        scaler.fit(searchspace.comp_rep_bounds)
+        # Create a composite scaler from parameter-wise scaler objects
+        mapping: dict[tuple[int, ...], InputTransform] = {}
+        for p in searchspace.parameters:
+            idxs = searchspace.get_comp_rep_parameter_indices(p.name)
+            factory = self._make_parameter_scaler_factory(p)
+            if factory is None:
+                continue
+            transformer = factory(len(idxs))
+            mapping[idxs] = transformer
+        scaler = ColumnTransformer(mapping)
+
+        # Fit the scaler to the parameter bounds
+        scaler.fit(to_tensor(searchspace.comp_rep_bounds))
 
         return scaler
 
     def _make_output_scaler(
         self, objective: Objective, measurements: pd.DataFrame
     ) -> OutcomeTransform | _NoTransform:
         """Make the output scaler for transforming computational dataframes."""
-        import torch
-
         scaler = self._make_target_scaler()
         if scaler is None:
             return _IDENTITY_TRANSFORM
 
         # TODO: Consider taking into account target boundaries when available
-        scaler(torch.from_numpy(objective.transform(measurements).values))
+        scaler(to_tensor(objective.transform(measurements)))
         scaler.eval()
 
         return scaler
 
-    def transform_inputs(self, df: pd.DataFrame, /) -> pd.DataFrame:
-        """Transform an experimental parameter dataframe."""
-        if self._input_transform is None:
-            raise ModelNotTrainedError("The model must be trained first.")
-        return self._input_transform(df)
+    def posterior(self, candidates: pd.DataFrame, /) -> Posterior:
+        """Compute the posterior for candidates in experimental representation.
+
+        Takes a dataframe of parameter configurations in **experimental representation**
+        and returns the corresponding posterior object. Therefore, the method serves as
+        the user-facing entry point for accessing model predictions.
+        """
+        return self._posterior_comp_rep(
+            to_tensor(self._searchspace.transform(candidates))
+        )
 
-    def transform_outputs(self, df: pd.DataFrame, /) -> pd.DataFrame:
-        """Transform an experimental measurement dataframe."""
-        if self._output_transform is None:
-            raise ModelNotTrainedError("The model must be trained first.")
-        return self._output_transform(df)
+    def _posterior_comp_rep(self, candidates: Tensor, /) -> Posterior:
```

Comment:
You mean why this is made available as a separate method in the first place? This is pretty much the reason for why the entire refactoring though process took so long: In the past, we've always tried to have only two "posterior" methods and integrate the mentioned one somehow into the other two, which always caused confusion and unclear interfaces. 

After thinking a bit about one what is actually required, I noticed that many problems are caused exactly by the fact that we try to implement functionality using only two interfaces even though the surrounding architecture really imposes three separate interfaces on us (see also PR description):
* There is the public dataframe-based posterior we want to expose on a user level
* There is the low-level mathematical posterior logic which we want subclasses to implement without having to bother about transforms and whatever 
* And what we always tried to ignore: there is the one that we need on the computational layer to be passed as callable to botorch. 

I've always tried to come up with a solution that does not require the last one, but I think this caused the pain in the previous versions and was always source of errors. Having all imposed interfaced defined makes the logic much easier, I think.

But please let my know if you see a better alternative, because this was really the tricky part in finding the design.

Reply:
Topic was discussed in meeting, so will close here. Feel free to reopen when further discussion is needed.